### DISCUSSION TITLE
Does the Validation prop need anything additional to function ?

### DISCUSSION THREAD

**Discussion by keerthankrishna** [react-querybuilder/react-querybuilder#441] [General]
I'm trying to setup some **validations** before the _QueryBuilder_ data is sent to the Server but for some reason even the ` defaultValidator ` is not working - What am i missing ?

Here's the CodeSandbox link: https://codesandbox.io/s/laughing-tom-wlkrup-react-qb-wlkrup?file=/src/App.js
It's just the basic react QueryBuilder skeleton code with ` defaultValidator ` added to it for validation

**Comment by jakeboone02** [comment]
The `defaultValidator` only checks for empty groups (unless you're using `independentCombinators`, which you're not). Also, the `validator` prop only affects what classes are added to the components. It doesn't actually change anything about the layout or styling unless you have some CSS that reacts to those classes.

For example, you could add this to your sandbox's `styles.css` to make it obvious when a group is empty:

```css
.queryBuilder-invalid {
  border-color: rebeccapurple;
  border-width: 4px;
}
```

That would give you something like this (until you add a rule to the "invalid" group):

![image](https://user-images.githubusercontent.com/366438/210899387-2c3b5972-4b88-4a13-916c-3405fc5c76e9.png)

Live example here: https://codesandbox.io/s/laughing-tom-wlkrup-react-qb-forked-2z9ucz?file=/src/App.js

If you want to run validation on the query object as you pass it through `formatQuery`, see [this documentation](https://react-querybuilder.js.org/docs/api/export#validation).

**Reply by keerthankrishna** [reply]
Perfect, i missed the addition of the style - thanks for clarifying that @jakeboone02! How did you add the styles which you show in this  [Demo](https://react-querybuilder.js.org/demo#addRuleToNewGroups=false&autoSelectField=false&autoSelectOperator=false&debugMode=false&disabled=false&enableDragAndDrop=false&independentCombinators=false&justifiedLayout=false&listsAsArrays=false&parseNumbers=false&resetOnFieldChange=false&resetOnOperatorChange=false&showBranches=false&showCloneButtons=false&showCombinatorsBetweenRules=false&showLockButtons=false&showNotToggle=false&validateQuery=true) ? (I see even the blank input field turns purple if left empty - what style are you adding to achieve this?)
![image](https://user-images.githubusercontent.com/24373382/210921500-733ec30c-6837-47ed-ade7-ccff86eb7f76.png)


Is it possible to call this _invalid style_ (` validator ` function) only upon click of a button ?
**Example** - _If I have a SAVE button, and I want the validation and invalid style to be added only when clicking the button, how do I do it ? Right now, I think this is preloaded into the QueryBuilder component as an additional prop which immediately gets called_

In addition to the above, I'm thinking of showing a message that the rule is not fully formed by using the below option (` formatQuery `) where I check the output but how do i change the default fallback expression by getting a _boolean_ value rather than _(1=1)_ for all the negative scenarios ? _(forgive me if this last question is too lame)_
![image](https://user-images.githubusercontent.com/24373382/210922730-e22ab7e5-d291-404d-86a3-4058751c0995.png)

CodeSandbox (forked from your last update) [here](https://codesandbox.io/s/laughing-tom-wlkrup-react-qb-forked-g3lqf7?file=/src/App.js)

**I think I did not understand how to place the "Validator" function correctly**

**Reply by jakeboone02** [reply]
One thing the docs don't make _super_ clear is that the demo implements field-specific validation in addition to the query-level `defaultValidator`.

I modified your sandbox to do the same, as well as all the other enhancements and fixes discussed below:

https://codesandbox.io/s/ojx2ts?file=/src/App.js

```js
const validator = r => !!r.value;
const fields = [
  { name: 'firstName', label: 'First Name', placeholder: 'Enter First Name', validator },
  { name: 'lastName', label: 'Last Name', placeholder: 'Enter Last Name', validator },
];
```

The `validator` property of a field is used to validate each rule where that field has been chosen. You can see the actual configuration for the demo by clicking the "Other files" section at the bottom of the page.

One way to implement a validate-on-save feature would be to apply a class to the wrapper element when the button is clicked. You can then remove the class when the query is updated (using `onQueryChange`) to remove the validation styles.

```jsx
export const App = () => {
  const [query, setQuery] = useState(initialQuery);
  const [validating, setValidating] = useState(false);

  const saveClick = () => {
    const output = formatQuery(query, {
      format: 'sql',
      validator: q => q.rules.length > 0,
    });
    console.log(output);
    setValidating(true); // <- turn validation on
  };

  return (
    <div className={validating ? 'validateQuery' /* <- applied only when validating */ : ''}>
      <QueryBuilder
        fields={fields}
        query={query}
        independentCombinators={true}
        onQueryChange={q => {
          setQuery(q);
          setValidating(false); // <- turn validation off
        }}
        validator={defaultValidator}
        controlElements={{ fieldSelector: MyFieldSelector }}
      />
// ...
```

The styles for the demo when validation is enabled look like this (compiled to CSS here; [SCSS source code here](https://github.com/react-querybuilder/react-querybuilder/blob/2d06d29fd36ed49778d1a420b41cf95c33067bdd/website/src/pages/demo/_styles/demo.scss)):

```scss
.validateQuery .queryBuilder .ruleGroup.queryBuilder-invalid {
  background-color: #66339966;
}
.validateQuery .queryBuilder .ruleGroup.queryBuilder-invalid .ruleGroup-addRule {
  font-weight: bold !important;
}
.validateQuery .queryBuilder .ruleGroup.queryBuilder-invalid > .ruleGroup-header::after {
  content: "Empty groups are considered invalid. Avoid them by using addRuleToNewGroups.";
  color: white;
}
.validateQuery .queryBuilder .rule.queryBuilder-invalid .rule-value {
  background-color: #66339966;
}
.validateQuery .queryBuilder .rule.queryBuilder-invalid .rule-value::placeholder {
  color: #47246b;
}
```

I also fixed an issue where you had a component defined within another component. See [documentation here](https://react-querybuilder.js.org/docs/tips/common-mistakes#custom-component-as-closure).

For `formatQuery`, the validator works at the query level. I implemented the following validator in the sandbox to add the logic from the field-level validators to the default validator logic:

```js
const formatQueryValidator = q => {
  const validations = {};

  const processGroup = g => {
    if (g.rules.length === 0) {
      validations[g.id] = false;
    }
    g.rules.forEach(r => {
      if ('rules' in r) {
        processGroup(r);
      } else {
        if (!r.value) {
          validations[r.id] = false;
        }
      }
    });
  };

  processGroup(q);

  return validations;
};
```

As for `fallbackExpression`, see [docs here](https://react-querybuilder.js.org/docs/api/export#fallback-expression). The thing is, `fallbackExpression` only applies when `formatQuery` can't figure out what else to do for the given format, like when there's an empty group and the format is "sql". SQL can't handle `()` as a condition, so the fallback is `(1 = 1)` to maintain valid syntax without affecting the logic (hopefully).

By default, `formatQuery` will completely ignore rules/groups that don't pass validation. If you want to change the output for rules/groups that don't pass validation, you'd probably need to provide a custom `ruleProcessor` ([docs here](https://react-querybuilder.js.org/docs/api/export#rule-processor)).

**Reply by keerthankrishna** [reply]
Thank you so much for the detailed explanation and making room for those enhancements as well as fixing the issues ...and showing the example. I will go through this so that I understand everything but your example is working just like how I wanted it to!
